iceoryx_hoofs/include/iceoryx_hoofs/internal/design_pattern/polymorphic_handler.inl

```diff
@@ -0,0 +1,163 @@
+
+// 2. If any thread changes the handler with set or reset, it will:
+//    - be detected by comparison with local handler
+//    - set the current handler to the new handler
+//
+
+// On any call after the handler was changed in another thread
+// 1. We detect the change by comparison with the localHandler (can assume that addresses are unique by
+// design as we use singletons in static memory)
+// 2. If it was changed we perform a synchronizing load and obtain the current handler
+// Note that it may change again but this is unavoidable if it can change concurrently
+// without locks and not a problem.
+template <typename Interface, typename Default, typename Hooks>
+Interface& PolymorphicHandler<Interface, Default, Hooks>::get() noexcept
+{
+    // NOLINTNEXTLINE(cppcoreguidelines-avoid-non-const-global-variables) false positive, thread_local
+    thread_local Interface* localHandler = getCurrent(); // initialized once per thread on first call
+
+    // weak load, required since it could be set in another thread concurrently
+    auto* currentHandler = instance().m_current.load(std::memory_order_relaxed);
+
+    // is the known localHandler outdated?
+    // NB: this is a variation of double checked locking but with atomics:
+    // we avoid the more costly operation on the hot path unless there was a change (rare case)
+    if (localHandler != currentHandler)
+    {
+        // stronger sync of memory
+        localHandler = getCurrent();
+        // note that it may concurrently change again but we do not loop to check as there is no
+        // point in such a lock-free code (it may also change while returning)
+    }
+
+    return *localHandler;
+}
+
+template <typename Interface, typename Default, typename Hooks>
+template <typename Handler>
+Interface* PolymorphicHandler<Interface, Default, Hooks>::set(StaticLifetimeGuard<Handler> handlerGuard) noexcept
+{
+    static_assert(std::is_base_of<Interface, Handler>::value, "Handler must inherit from Interface");
+    static StaticLifetimeGuard<Handler> guard(handlerGuard);
+    // we now have protected the handler instance and it will exist long enough
+    return setHandler(StaticLifetimeGuard<Handler>::instance());
+}
+
+template <typename Interface, typename Default, typename Hooks>
+Interface* PolymorphicHandler<Interface, Default, Hooks>::setHandler(Interface& handler) noexcept
+{
+    auto& ins = instance();
+    // m_current is now guaranteed to be set
+
+    if (ins.m_isFinal.load(std::memory_order_acquire))
+    {
+        // it must be ensured that the handlers still exist and are thread-safe,
+        // this is ensured for the default handler by m_defaultGuard
+        // (the primary guard that is constructed with the instance alone is not sufficient)
+        Hooks::onSetAfterFinalize(*getCurrent(), handler);
+        return nullptr;
+    }
+
+    auto prev = ins.m_current.exchange(&handler, std::memory_order_acq_rel);
```

Comment:
This is inherently racy. Another thread could call `finalize` right after the check above. This is okay but should be documented.

Reply:
True, but calling some function and it taking effect are two things. Of course, taking effect here should mean the flag is set. However, it is not a problem if the `setHandler` function stated earlier, it is still ok to take effect and has well-defined semantics.